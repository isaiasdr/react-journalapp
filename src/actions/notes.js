import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, updateDoc } from "firebase/firestore";
import { db } from "../firebase/firebase-config";

import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from '../helpers/fileUpload';


export const startNewNote = () => {
    return async ( dispatch, getState ) => {
        const uid = getState().auth.uid;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        };

        const docRef = await addDoc(collection(db, `${ uid }/journal/notes`), newNote);
        dispatch( activeNote( docRef.id, newNote ) );
        dispatch( addNewNote( docRef.id, newNote ) );
    }
}

export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = ( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
});

export const startLoadingNotes = ( uid ) => {
    return async ( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch ( setNotes( notes ) );
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        if (!note.url)
            delete note.url;

        const noteToFiretore = { ...note };
        delete noteToFiretore.id;
        
        const docRef = await updateDoc(doc(db, `${ uid }/journal/notes/${ note.id }`), noteToFiretore);
        dispatch ( refreshNote( note.id, note ) );
        Swal.fire('Saved!', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
});

export const startUploadingImage = ( file ) => {
    return async ( dispatch, getState ) => {
        const { active:activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload( file );
        activeNote.url = fileUrl;

        dispatch( startSaveNote( activeNote ) );
        Swal.close();
    }
}

export const startDeleting = ( id ) => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const result = await deleteDoc( doc(db, `${ uid }/journal/notes/${ id }`) );
        dispatch( deleteNote( id ) );
        Swal.fire('Deleted!', 'Your note has been deleted.', 'success');
    }
}

export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
});

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
});