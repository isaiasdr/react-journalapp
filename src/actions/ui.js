import Swal from 'sweetalert2';

import { types } from "../types/types"


export const setError = ( err ) => {
    Swal.fire({
        title: 'Error',
        text: err,
        icon: 'error'
    });

    return {
        type: types.uiSetError,
        payload: err
    }
};

export const removeError = () => ({
    type: types.uiRemoveError
});

export const startLoading = () => ({
    type: types.uiStartLoading
});

export const finishLoading = () => ({
    type: types.uiFinishLoading
});