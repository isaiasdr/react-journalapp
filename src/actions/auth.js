import Swal from 'sweetalert2';
import { getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from "firebase/auth";

import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { startLoading, finishLoading } from "./ui";
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch( startLoading() );

        const auth = getAuth();

        signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                dispatch( login(user.uid, user.displayName) );
                dispatch( finishLoading() );
            })
            .catch( (e) => {
                //console.log(e);
                dispatch( finishLoading() );
                Swal.fire({
                    title: 'Error',
                    text: e.message,
                    icon: 'error'
                });
            });
    }
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        dispatch( startLoading() );
        const auth = getAuth();

        createUserWithEmailAndPassword(auth, email, password)
            .then( async ({user}) => {
                await updateProfile(auth.currentUser, { displayName: name });

                dispatch( login(user.uid, user.displayName) );
                dispatch( finishLoading() );
            })
            .catch( (e) => {
                //console.log(e);
                dispatch( finishLoading() );
                Swal.fire({
                    title: 'Error',
                    text: e.message,
                    icon: 'error'
                });
            });
    }
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();

        signInWithPopup(auth, googleAuthProvider)
            .then( ({user}) => {
                dispatch( login(user.uid, user.displayName) );
            });
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        const auth = getAuth();

        await signOut(auth)
            .then( dispatch( logout() ) );

        dispatch( noteLogout() );
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});