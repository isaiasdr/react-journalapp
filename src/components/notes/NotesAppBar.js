import React from 'react'
import { useSelector, useDispatch } from 'react-redux';

import { startSaveNote } from '../../actions/notes';


export const NotesAppBar = () => {

    const { active } = useSelector(state => state.notes);
    const dispatch = useDispatch();

    const handleSave = () => {
        dispatch( startSaveNote(active) );
    }

    return (
        <div className='notes__appbar'>
            <span>07 de Septiembre 2022</span>

            <div>
                <button className='btn'>
                    Picture
                </button>

                <button
                    className='btn'
                    onClick={ handleSave }
                >
                    Save
                </button>
            </div>
        </div>
    )
}
