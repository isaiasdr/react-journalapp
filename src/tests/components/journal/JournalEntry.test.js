import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";

import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

const note = {
    id: '12345',
    title: 'Test Note',
    body: 'This is a test note',
    date: new Date().getTime()
};

let store = mockStore(initState);

store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <JournalEntry {...note} />
    </Provider>
);

describe('Tests in <JournalEntry />', () => {
    test('should show correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call the activeNote action', () => {
        wrapper.find('.journal__entry').simulate('click');

        expect( store.dispatch ).toHaveBeenCalledWith( activeNote(note.id,  note) );
    });
});
