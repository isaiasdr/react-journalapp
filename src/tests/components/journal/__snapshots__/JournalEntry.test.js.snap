// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry /> should show correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Test Note
    </p>
    <p
      className="journal__entry-content"
    >
      This is a test note
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Sunday
    </span>
    <h4>
      9th
    </h4>
  </div>
</div>
`;
